libfwupdplugin/fu-udev-device.c

```diff
@@ -1552,6 +1628,110 @@ fu_udev_device_write_sysfs(FuUdevDevice *self,
 				       error);
 }
 
+/**
+ * fu_udev_device_write_sysfs_byte_array:
+ * @self: a #FuUdevDevice
+ * @attr: sysfs attribute name
+ * @buf: data to write into the attribute
+ * @timeout_ms: IO timeout in milliseconds
+ * @error: (nullable): optional return location for an error
+ *
+ * Writes raw data into a sysfs attribute.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 2.0.1
+ **/
+gboolean
+fu_udev_device_write_sysfs_byte_array(FuUdevDevice *self,
+				      const gchar *attr,
+				      GByteArray *buf,
+				      guint timeout_ms,
+				      GError **error)
+{
+	g_autofree gchar *path = NULL;
+	g_autoptr(FuIOChannel) io_channel = NULL;
+
+	g_return_val_if_fail(FU_IS_UDEV_DEVICE(self), FALSE);
+	g_return_val_if_fail(attr != NULL, FALSE);
+	g_return_val_if_fail(buf != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	/* emulated */
+	if (fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_EMULATED))
+		return TRUE;
```

Comment:
something I wonder mostly in terms of us having to be ABI stable again is if we're ever going to want to have validation that the write wrote all the correct data.

Reply:
Good idea! Fixed in 1597d6bc92d7e1aac8ac4de9cd7d6665b8f4be51